Catch Send/Receive errors and make them sync errors

rmunn opened the pull request:
Fixes #1888.

rmunn commented:
Hmm, this is tricky; Chorus only sends certain messages to the `progress`, and our SafeLoggingProgress does (as Code Rabbit pointed out) end up causing the LfMergeBridgeResult object to have an empty string in `progress`. I'll need to revamp the S/R service and helpers a little to ensure that success or failure is caught earlier.

hahn-kev commented on backend/FwHeadless/Services/CombiningProgress.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace FwHeadless.Services;
+
+public class CombiningProgress(params IEnumerable<IProgress?> progresses) : IProgress
```
let's use `ICollection<T>` instead. When  consuming an `IEnumerable` it's not safe to assume that you can iterate it more than once

rmunn commented on backend/FwHeadless/Services/SendReceiveHelpers.cs:
```diff
@@ -19,15 +21,23 @@ public SendReceiveAuth(FwHeadlessConfig config) : this(config.LexboxUsername, co
 
     public record LfMergeBridgeResult(string Output, string ProgressMessages);
 
+    public static bool IsSuccess(LfMergeBridgeResult result)
```
Done in commit d222c025 which made the code quite simpler; thanks for the suggestion. I switched it to ErrorEncountered because the consuming code was needing to do `if (!IsSuccess)` and the underlying IProgress uses ErrorEncountered anyway.

rmunn commented on backend/FwHeadless/Services/SendReceiveHelpers.cs:
```diff
@@ -19,15 +21,23 @@ public SendReceiveAuth(FwHeadlessConfig config) : this(config.LexboxUsername, co
+    public static bool IsSuccess(LfMergeBridgeResult result)
+    {
+        if (result.ProgressMessages.Contains("Unable to complete the send/receive.")) return false;
+        if (result.ProgressMessages.Contains("Rolling back...")) return false;
```
Now that I've verified that the ErrorEncountered property is correctly set when Chorus calls WriteError (or WriteException, in turn calls WriteError), I can just use that and stop parsing output. It's also not safe to search for such generic strings because the word "failure" or "error" could easily appear in filenames (audio files, for example, are often named with the English-language gloss of the vernacular word) so it would be prone to false positives. (It's safe in our tests because our test data doesn't include those words in filenames). So now that I'm using ErrorEncountered I've gotten rid of the text parsing.

hahn-kev commented on backend/FwHeadless/Services/SendReceiveHelpers.cs:
```diff
@@ -17,17 +19,21 @@ public record SendReceiveAuth(string Username, string Password)
     };
 
-    public record LfMergeBridgeResult(string Output, string ProgressMessages);
+    public record LfMergeBridgeResult(string Output, IProgress progress)
```
I didn't think of this when I made my suggestion. But this will actually make `IProgress` a property of LfMergeBridgeResult, and if it's serialized then it will include that object, which we probably done want. I'd suggest this:
```C#
public record LfMergeBridgeResult(string Output, bool ErrorEncountered)
{
    public LfMergeBridgeResult(string output, IProgress progress): this(output, progress.ErrorEncountered);
}
```
be defining another constructor we can achieve the same result, but without adding `IProgress` as a property of the record.

hahn-kev reviewed (changes requested):
Looks good, one minor change, I should have realized it was a problem before I made the suggestion.

rmunn commented:
@hahn-kev - Addressed your comment; thanks for the catch. Please see https://github.com/sillsdev/languageforge-lexbox/pull/1924#discussion_r2308887873 and https://github.com/sillsdev/languageforge-lexbox/pull/1924#discussion_r2308890729 and let me know whether commit 2bfe5759 is necessary.

hahn-kev commented:
@rmunn I'm fine either way, I probably prefer the simpler option, I think it would be weird to create the result object and then the result changes after the fact thought.
